File: android/app/src/main/java/org/evergreenlabs/mytwitter/data/Post.kt
package org.evergreenlabs.mytwitter.data

import com.google.firebase.firestore.DocumentId
import com.google.firebase.firestore.Exclude
import com.google.firebase.firestore.IgnoreExtraProperties
import java.util.Date

@IgnoreExtraProperties
data class MediaItem(
    val type: String? = null,
    val url: String? = null,
    val previewUrl: String? = null,
    val videoUrl: String? = null,
    val width: Int? = null,
    val height: Int? = null,
    val alt: String? = null,
) {
    @get:Exclude
    val isVideo: Boolean
        get() {
            val t = (type ?: "").lowercase()
            return t == "video" || t == "animated_gif"
        }

    @get:Exclude
    val displayImageUrl: String?
        get() = previewUrl?.takeIf { it.isNotBlank() } ?: url?.takeIf { it.isNotBlank() }

    @get:Exclude
    val playableVideoUrl: String?
        get() = videoUrl?.takeIf { it.isNotBlank() }

    /** Width/height from X when present; otherwise a 16:9 fallback so layouts don't stretch. */
    @get:Exclude
    val aspectRatio: Float
        get() {
            val w = (width ?: 0).toFloat()
            val h = (height ?: 0).toFloat()
            if (w <= 0f || h <= 0f) return 16f / 9f
            return (w / h).coerceIn(0.45f, 2.4f)
        }
}

@IgnoreExtraProperties
data class LinkPreview(
    val url: String? = null,
    val tcoUrl: String? = null,
    val expandedUrl: String? = null,
    val displayUrl: String? = null,
    val domain: String? = null,
    val title: String? = null,
    val description: String? = null,
    val imageUrl: String? = null,
) {
    @get:Exclude
    val openUrl: String?
        get() = expandedUrl?.takeIf { it.isNotBlank() }
            ?: url?.takeIf { it.isNotBlank() }
            ?: tcoUrl?.takeIf { it.isNotBlank() }
}

@IgnoreExtraProperties
data class Post(
    @DocumentId val id: String? = null,
    val text: String? = null,
    val url: String? = null,
    val authorId: String? = null,
    val authorHandle: String? = null,
    val authorName: String? = null,
    val authorAvatar: String? = null,
    val createdAt: Date? = null,
    val isRetweet: Boolean? = null,
    val repostedById: String? = null,
    val repostedByHandle: String? = null,
    val repostedByName: String? = null,
    val repostedByAvatar: String? = null,
    val media: List<MediaItem>? = null,
    val mediaUrls: List<String>? = null,
    val linkPreview: LinkPreview? = null,
) {
    @get:Exclude
    val tweetId: String get() = id.orEmpty()

    @get:Exclude
    val displayHandle: String
        get() {
            val h = authorHandle?.trim().orEmpty()
            return h.ifEmpty { "unknown" }
        }

    @get:Exclude
    val displayName: String
        get() {
            val n = authorName?.trim().orEmpty()
            return n.ifEmpty { "@$displayHandle" }
        }

    @get:Exclude
    val xUrl: String
        get() = url?.takeIf { it.isNotBlank() } ?: "https://x.com/i/status/$tweetId"

    @get:Exclude
    val resolvedMedia: List<MediaItem>
        get() {
            if (!media.isNullOrEmpty()) return media
            return (mediaUrls ?: emptyList()).mapNotNull { u ->
                if (u.isBlank()) null
                else MediaItem(type = "photo", url = u, previewUrl = u)
            }
        }

    @get:Exclude
    val showAsRepost: Boolean
        get() = isRetweet == true &&
            !repostedByHandle.isNullOrBlank() &&
            repostedById != authorId
}

File: ios/MyTwitter/Models/Post.swift
import Foundation
import FirebaseFirestore

struct MediaItem: Codable, Hashable, Identifiable {
    var id: String {
        "\(type ?? "media")-\(url ?? previewUrl ?? videoUrl ?? "unknown")"
    }
    var type: String?
    var url: String?
    var previewUrl: String?
    var videoUrl: String?
    var width: Int?
    var height: Int?
    var alt: String?

    var isVideo: Bool {
        let t = (type ?? "").lowercased()
        return t == "video" || t == "animated_gif"
    }

    var displayImageURL: URL? {
        URL(string: previewUrl ?? url ?? "")
    }

    var playableVideoURL: URL? {
        guard let videoUrl, !videoUrl.isEmpty else { return nil }
        return URL(string: videoUrl)
    }

    /// Width/height from X when present; otherwise a 16:9 fallback so layouts don't stretch.
    var aspectRatio: CGFloat {
        let w = CGFloat(width ?? 0)
        let h = CGFloat(height ?? 0)
        guard w > 0, h > 0 else { return 16.0 / 9.0 }
        return min(max(w / h, 0.45), 2.4)
    }
}

struct LinkPreview: Codable, Hashable {
    var url: String?
    var tcoUrl: String?
    var expandedUrl: String?
    var displayUrl: String?
    var domain: String?
    var title: String?
    var description: String?
    var imageUrl: String?

    var openURL: URL? {
        URL(string: expandedUrl ?? url ?? tcoUrl ?? "")
    }
}

struct Post: Codable, Identifiable, Hashable {
    @DocumentID var id: String?
    var text: String?
    var url: String?
    var authorId: String?
    var authorHandle: String?
    var authorName: String?
    var authorAvatar: String?
    var createdAt: Date?
    var isRetweet: Bool?
    var repostedById: String?
    var repostedByHandle: String?
    var repostedByName: String?
    var repostedByAvatar: String?
    var media: [MediaItem]?
    var mediaUrls: [String]?
    var linkPreview: LinkPreview?

    var tweetId: String { id ?? "" }

    var displayHandle: String {
        let h = (authorHandle ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        return h.isEmpty ? "unknown" : h
    }

    var displayName: String {
        let n = (authorName ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        return n.isEmpty ? "@\(displayHandle)" : n
    }

    var xURL: URL {
        if let url, let parsed = URL(string: url) { return parsed }
        return URL(string: "https://x.com/i/status/\(tweetId)")!
    }

    var resolvedMedia: [MediaItem] {
        if let media, !media.isEmpty { return media }
        return (mediaUrls ?? []).compactMap { u in
            guard !u.isEmpty else { return nil }
            return MediaItem(type: "photo", url: u, previewUrl: u, videoUrl: nil, width: nil, height: nil, alt: nil)
        }
    }

    var showAsRepost: Bool {
        (isRetweet == true) && !(repostedByHandle ?? "").isEmpty
            && (repostedById != authorId)
    }
}
